File: 07/00/DayofYearSet.java
import java.util.ArrayList;

/**
 * DayofYearSet
 *
 * @author s.burak
 * @version 1.0.0
 * @since  2022-05-01
 */

public class DayofYearSet{
    
    static class DayOfYear{
        
        private int day;
        private int month;

        public DayOfYear(){
            day = 1;
            month = 1;
        }

        public DayOfYear(int d,int m){
            int numberOfDay[] = {31,28,31,30,31,30,31,31,30,31,30,31};

            if(m<=0 || m >12){
                //ERROR
                month = 1;
                day = 1;
                return;
            }

            else if(d<=0 || d>numberOfDay[m-1]){
                //ERROR
                month = m;
                day = 1;
                return;
            }

            else {
                month = m;
                day = d;
                return;
            }
                

        }

        public void setDay(int d){
            int numberOfDay[] = {31,28,31,30,31,30,31,31,30,31,30,31};
            
            if(d<=0 || d>numberOfDay[month-1]){
                //ERROR
                return;
            }
            
            else day = d;
        }

        public void setMonth(int m){
            
            if(m<=0 || m >12){
                //ERROR
                return;

            }

            else month = m;

        }

        public void setAll(int d,int m){
            
            int numberOfDay[] = {31,28,31,30,31,30,31,31,30,31,30,31};

            if(m<=0 || m >12){
                return;
            }

            else if(d<=0 || d>numberOfDay[m-1]){
                return;
            }

            else {
                month = m;
                day = d;
                return;
            }

        }

        public int getDay(){
            return day;
        }
        
        public int getMonth(){
            return month;
        }


    }

    /** Constructs an empty set */

    public DayofYearSet(){
        setSize = 0;
    }

    /** 
    * Constructs a set containing the elements of the ArrayList
    * @param dayOfYearList ArrayList of DayofYear */

    public DayofYearSet(ArrayList<DayofYearSet.DayOfYear> dayOfYearList){
        setSize = 0;

        for(int i = 0;i<dayOfYearList.size();i++){
            add(dayOfYearList.get(i));
        }

    }

    /** 
    *Adds an element to the set. 
    * @param element to be added to the set
    */

    public void add(DayOfYear element){
        
        if(isThere(element))
            return;
        
        DayOfYear temp[] = new DayOfYear[size()+1];

        for(int i=0;i<size();i++){
            temp[i] = setArr[i];
        }

        temp[size()] = element;
        
        setArr = temp;

        setSize++;

        liveDays++;
        
    }

    /** Removes last element from the set. */

    public void remove(){
        
        DayOfYear temp[] = new DayOfYear[size()-1];
        
        for(int i=0;i<size()-1;i++){
            temp[i] = setArr[i];
        }
    
        setArr = temp;

        setSize--;

        liveDays--;
        
    }

    /**
    * Removes the element at the specified position in this set.
    * @param index of the element to be removed
     */
    public void remove(int index){
        
        DayOfYear temp[] = new DayOfYear[size()-1];
        
        for(int i=0,j=0;i<size();i++,j++){
            
            if(i==index){
                j--;
                continue;
            }

            temp[j] = setArr[i];
        }
    
        setArr = temp;

        setSize--;

        liveDays--;
        
    }

    /** 
    * Prints DayofYearSet details. 
    * @return string representation of this set*/

    public String toString(){
        
        if(size() ==0)
        return "Set is empty\n";
        
        String strOutput = "";

        for(int i = 0; i<size();i++){
            strOutput += "Element " + i + "  ->   Day:" + setArr[i].getDay() + "   Month:"+ setArr[i].getMonth()+"\n";
        }

        return strOutput;
    }

    /** 
    * Takes another DayofYearSet and compares with this. 
    * @param obj other set to compare with this set
    * @return true if their elements are equal regardless of the keeping order ,otherwise false*/

    public boolean equals(Object obj){
        
        if(obj==this)
            return true;
        
        if((obj.getClass() != this.getClass()))
            return false;
        
        DayofYearSet other = (DayofYearSet) obj;


        for(int i=0;i<size();i++){
            if(!other.isThere(setArr[i])){
                return false;
            }
        }
        return true;
    }

    /** 
    * Returns the complement of this set
    * @return complement of set*/

    public DayofYearSet complement(){
        int numberOfDay[] = {31,28,31,30,31,30,31,31,30,31,30,31};

        DayofYearSet tempSet = new DayofYearSet();
        
        
        
        for(int i=0 ; i<12;i++){
            for(int j = 0; j<numberOfDay[i];j++){
                DayOfYear tempDay = new DayOfYear(j+1,i+1);
                
                if(!isThere(tempDay))
                    tempSet.add(tempDay);
            }
        }

        return tempSet;
    }

    /** 
    * Takes another DayofYearSet and returns the union of these two sets
    * @param other the set to be union with this set 
    * @return union of two set
    
    
     */
    public DayofYearSet union(DayofYearSet other){
        
        DayofYearSet temp = new DayofYearSet();

        if(empty()){
            return other.copy();
        }

        else if(other.empty()){
            return copy();
        }

        else{
            
            for(int i = 0; i<setSize;i++){
                temp.add(setArr[i]);
            }

            for(int i = 0; i<other.size();i++){
                temp.add(other.setArr[i]);
            }

            return temp;
        }
    }

    /** 
    *Takes another DayofYearSet and returns the intesection of these two sets
    * @param other the set to be intersection with this set 
    * @return intersection of two set*/
    public DayofYearSet intersection(DayofYearSet other){
        
        DayofYearSet temp = new DayofYearSet();

        if(empty() || other.empty()){
            return temp;
        }

        else{
            
            for(int i = 0; i<setSize ; i++){
                if(other.isThere(setArr[i]))
                    temp.add(setArr[i]);
            }
            return temp;
        }
    }

    /** 
    *Takes another DayofYearSet and returns the difference of these two sets 
    * @param other the set to be difference with this set 
    * @return difference of two set*/


    public DayofYearSet difference(DayofYearSet other){
        
        DayofYearSet temp = new DayofYearSet();

        if(empty() || other.empty()){
            return copy();
        }
        
        else{
            
            for(int i = 0; i<setSize ; i++){
                if(!other.isThere(setArr[i]))
                    temp.add(setArr[i]);
            }
            
            return temp;
        }
    
    }

    private DayofYearSet copy(){
        
        DayofYearSet tempSet = new DayofYearSet();
        
        for(int i=0;i<size();i++){
            tempSet.add(setArr[i]);
        }

        return tempSet;
    }

    /** 
    * Returns the number of elements. 
    * @return number of elements */

    public int size(){
        return setSize;
    }

    /** Checks if set is empty. 
    * @return true if there is no element in the set ,otherwise false*/

    public boolean empty(){
        if(setSize == 0)
            return true;
        else return false;
    }

    protected void finalize(){
        liveDays = liveDays - size();
    }


    private boolean isThere(DayOfYear element){
        for(int i=0;i<size();i++){
            if(setArr[i].getDay() == element.getDay() && setArr[i].getMonth() == element.getMonth()){
                return true;
            }
        }

        return false;
    }

    /** 
    *returns the total number of DayOfYear objects alive in all the sets. 
    *@return the total number of DayOfYear objects alive in all the sets. */

    public static int numberOfLiveDay(){
        return liveDays;
    }

    public static int liveDays = 0;
    private DayOfYear[] setArr;
    private int setSize;
    
}

File: 04/src/dayofyearset.cpp
#include <iostream>
#include <vector>
#include "dayofyearset.h"

using namespace std;
namespace sBurakYasar{

    int DayofYearSet :: totalDoy = 0;

    /*Default argument constructor*/
    DayofYearSet :: DayofYear :: DayofYear(int d,int m ){
        
        int const arrMonth[12] = {31,28,31,30,31,30,31,31,30,31,30,31}; //number of days in months

        if(m<0 || m>12){ //checks if input is valid
            cout << endl << "Invalid month vaule" << endl;
            exit(1); 
        }

        if(d < 0 || d > arrMonth[m-1]){ //checks if input is valid
            cout << endl << "Invalid day vaule" << endl;
            exit(1); 
        }

        day = d;
        month = m;
    }

    void DayofYearSet :: DayofYear :: setDay(int d){
        
        int const arrMonth[12] = {31,28,31,30,31,30,31,31,30,31,30,31}; //number of days in months
        
        if(d < 0 || d > arrMonth[month-1]){ //checks if input is valid
            cout << endl << "Invalid day vaule" << endl;
            exit(1); 
        }

        day = d;
    }

    void DayofYearSet :: DayofYear :: setMonth(int m){

        if(m<0 || m>12){  //checks if input is valid
            cout << endl << "Invalid month vaule" << endl;
            exit(1); 
        }

        month = m;
    }

    void DayofYearSet :: DayofYear :: setAll(int d,int m){
        
        int const arrMonth[12] = {31,28,31,30,31,30,31,31,30,31,30,31}; //number of days in months
        
        

        if(m<0 || m>12){ //checks if input is valid
            cout << endl << "Invalid month vaule" << endl;
            exit(1); 
        }

        if(d < 0 || d > arrMonth[m-1]){ //checks if input is valid
            cout << endl << "Invalid day vaule" << endl;
            exit(1); 
        }

        day = d;
        month = m;

    }
    
    /*Getters*/
    int DayofYearSet :: DayofYear :: getDay()const{return day;} 
    int DayofYearSet :: DayofYear :: getMonth()const{return month;}

    bool DayofYearSet :: DayofYear :: operator ==(const DayofYear& other){ //checks if dayofyears are same
        if(day == other.day && month == other.month) return true;
        else return false;
    }

    ostream& operator<<(ostream &out,const DayofYearSet :: DayofYear& input){ //ostream operator overloading for dayofyear
        out << endl << "Day:" << input.getDay() << endl <<  "Month:" << input.getMonth() << endl;
        return out;
    }

    /*Constructor*/
    DayofYearSet :: DayofYearSet(){
        daySet = nullptr;
        setSize = 0; 
    }

    DayofYearSet :: DayofYearSet(const vector<DayofYear> &input){
        
        setSize = 0;

        if(input.empty()){
            daySet = nullptr;    
        }

        else{
            for(int i=0 ; i< input.size(); i++){
                add(input[i]); 
            }            
        }
    }

    DayofYearSet :: ~DayofYearSet(){
        
        if(!(daySet == nullptr))
        delete []daySet;
        totalDoy = totalDoy - setSize;
    }

    DayofYearSet :: DayofYearSet(const DayofYearSet& other){
        
        if(other.empty()){
            daySet = nullptr;
            setSize = 0;
        }

        else{
            
            daySet = new DayofYear[other.size()];
                
            for(int i=0;i<other.size();i++){
                daySet[i] = other[i];
            }
            setSize = other.size();
            totalDoy +=setSize;
        }

    }

    const DayofYearSet DayofYearSet :: operator=(const DayofYearSet& other){
        
        if(other.empty()){
            daySet = nullptr;
            setSize = 0;
        }

        else{
            
            daySet = new DayofYear[other.size()];
                
            for(int i=0;i<other.size();i++){
                daySet[i] = other[i];
            }
            setSize = other.size();
            totalDoy +=setSize;
        }

        return *this;
    }

    int DayofYearSet :: size()const{return setSize;}

    void DayofYearSet :: add(const DayofYear& input){
        
        if(empty()){
            daySet = new DayofYear[1];
            daySet[0] = input;
            setSize++;
            totalDoy++;
        }

        else if(!isThere(input)){
            
            DayofYear * temp = new DayofYear[setSize+1];
            
            for(int i=0;i<setSize;i++){
                temp[i] = daySet[i];
            }

            temp[setSize] = input;

            delete[] daySet;

            daySet = temp;

            
            setSize++;
            totalDoy++;
        }
        
    }

    void DayofYearSet :: remove(){
        if(empty()){
            cout << endl << "---Error:there is not any element in set ---" << endl;
            return;
        }

        else{
            
            if(setSize == 1){
                delete[] daySet;
                daySet = nullptr;
                setSize--;
                totalDoy--;
                return;
            }

            DayofYear * temp = new DayofYear[setSize-1];
            
            for(int i=0;i<setSize-1;i++){
                temp[i] = daySet[i];
            }

            delete[] daySet;

            daySet = temp;
            
            setSize--;
            totalDoy--;
        }
    }

    void DayofYearSet :: remove(int index){
        
        if(empty()){
            cout << endl << "---Error:there is not any element in set ---" << endl;
        }

        else{
            
            if(index>setSize){
                cerr << endl <<  "---Error:Invalid index---" << endl;
                return;
            }

            if(setSize == 1){
                delete[] daySet;
                daySet = nullptr;
                setSize--;
                totalDoy--;
                return;
            }

            DayofYear * temp = new DayofYear[setSize-1];
            int j = 0; 
            for(decltype(j)i=0;i<setSize-1;i++,j++){
                
                if(i==index) j++;

                temp[i] = daySet[j];
            }

            delete[] daySet;

            daySet = temp;
            
            setSize--;
            totalDoy--;
        }
    }

    void DayofYearSet :: remove(const DayofYear& input){
        
        if(empty()){
            cout << endl << "---ERROR:there is not any element in set---" << endl;
        }

        else{
            
            if(!isThere(input)){
                cout << endl << "this element is not in the set" << endl;
                return;
            }

            if(setSize == 1){
                delete[] daySet;
                daySet = nullptr;
                setSize--;
                totalDoy--;
                return;
            }

            DayofYear * temp = new DayofYear[setSize-1];
            int j = 0; 
            for(int i=0;i<setSize-1;i++,j++){
                
                if(daySet[i]==input) j++;

                temp[i] = daySet[j];
            }

            delete[] daySet;

            daySet = temp;
            
            setSize--;
            totalDoy--;
        }
        

    }

    bool DayofYearSet :: empty()const{
        if(setSize == 0 || daySet == nullptr)
            return true;
        else return false;
    }

    bool DayofYearSet :: isThere(const DayofYear& input)const{

        for(int i=0;i<setSize;i++){
            if(daySet[i] == input) return true;
        }
        return false;
    }

    ostream& operator<<(ostream &out,const DayofYearSet& input){
        
        //out << endl;

        if(input.empty()){
            out << "There is not any element in this set" << endl;
            return out;
        }

        for(int i=0;i<input.setSize ;i++){
            out << "Element "<< i+1 << "   ->\tDay:" << input.daySet[i].getDay() << "\tMonth:" << input.daySet[i].getMonth() << endl;
        }
        return out;
    }
    
    bool DayofYearSet :: operator==(const DayofYearSet& other)const{
        
        if(setSize != other.size())
            return false;


        for(int i = 0; i < other.size();i++){
            if(!isThere(other[i]) ){return false;}
        }
        return true;
    }

    bool DayofYearSet :: operator!=(const DayofYearSet& other)const{
        
        if( *this == other) return false;
        else return true;
    }

    const DayofYearSet DayofYearSet :: operator-(const DayofYearSet& other)const{
        
        DayofYearSet temp;

        if(empty() || other.empty()){
            return *this;
        }
        
        else{
            for(int i = 0; i<setSize ; i++){
                if(!other.isThere(daySet[i]))
                    temp.add(daySet[i]);
            }
            return temp;
        }
    }

    const DayofYearSet DayofYearSet :: operator+(const DayofYearSet& other)const{
        
        DayofYearSet temp;

        if(empty()){
            return other;
        }

        else if(other.empty()){
            return *this;
        }

        else{
            
            for(int i = 0; i<setSize;i++){
                temp.add(daySet[i]);
            }

            for(int i = 0; i<other.size();i++){
                temp.add(other[i]);
            }

            return temp;
        }
    }

    const DayofYearSet DayofYearSet :: operator^(const DayofYearSet& other)const{
        
        DayofYearSet temp;

        if(empty() || other.empty()){
            return temp;
        }

        else{
            
            for(int i = 0; i<setSize ; i++){
                if(other.isThere(daySet[i]))
                    temp.add(daySet[i]);
            }
            return temp;
        }
    }

    const DayofYearSet DayofYearSet :: operator!()const{
        int const arrMonth[12] = {31,28,31,30,31,30,31,31,30,31,30,31};

        DayofYearSet tempSet;
        DayofYear tempDay;
        
        
        for(int i=0 ; i<12;i++){
            for(auto j = 0; j<arrMonth[i];j++){
                tempDay.setAll(j+1,i+1);
                if(!isThere(tempDay))
                    tempSet.add(tempDay);
            }
        }

        return tempSet;
    }

    DayofYearSet :: DayofYear& DayofYearSet :: operator[](int index)const{
        
        //if(index < 0 || index>=setSize){cerr << endl << "---Error: Invalid Selection---" << endl; return  }

        
        return daySet[index];
    }

    int DayofYearSet :: getTotalDoy(){
        return totalDoy;
    }

}